crates/full-node/sov-sequencer/src/preferred/db/mod.rs

```diff
@@ -376,196 +375,52 @@ impl From<BatchToStore> for StoredBlob {
-            );
-
-            let batch_to_store = BatchToStore {
-                sequence_number,
-                blob_id,
-                visible_slot_number_after_increase,
-                visible_slots_to_advance,
-            };
-            backend.begin_rollup_block(batch_to_store).await?;
-        }
-
-        Ok(sequence_number)
-    }
+    ) -> Result<SequenceNumber>;
 
-    #[tracing::instrument(skip_all, level = "info")]
-    pub(crate) async fn insert_proof_blob(
+    async fn insert_proof_blob(
         &mut self,
         blob_id: BlobInternalId,
         data: Arc<[u8]>,
         sequence_number: SequenceNumber,
-    ) -> anyhow::Result<SequenceNumber> {
-        if let Some(backend) = &mut self.backend {
-            backend
-                .add_proof_blob(sequence_number, blob_id, data.clone())
-                .await?;
-        }
-
-        Ok(sequence_number)
-    }
-
-    #[tracing::instrument(skip_all, level = "info")]
-    pub(crate) async fn terminate_batch(&mut self, batch: BatchToStore) -> anyhow::Result<()> {
-        if let Some(backend) = &mut self.backend {
-            backend.end_rollup_block(batch).await?;
-            Self::debug_assert_in_progress_batch_is_none(
-                "Backend didn't remove in-progress batch from database when ending rollup block",
-                backend,
-                &self.shutdown_sender,
-            )
-            .await;
-        }
+    ) -> Result<SequenceNumber>;
 
-        Ok(())
-    }
+    async fn terminate_batch(&mut self, batch: BatchToStore) -> Result<()>;
 
-    #[tracing::instrument(skip_all, level = "info")]
-    pub(super) async fn prune_db(
-        &mut self,
-        prune_up_to_including: SequenceNumber,
-    ) -> anyhow::Result<()> {
-        if let Some(backend) = &mut self.backend {
-            if let Some(prune_up_to_including) = prune_up_to_including.checked_sub(PRUNING_LAG) {
-                backend.prune(prune_up_to_including).await?;
-            }
-        }
-        Ok(())
-    }
+    async fn prune_db(&mut self, prune_up_to_including: SequenceNumber) -> Result<()>;
+}
 
-    async fn debug_assert_in_progress_batch_is_none(
-        msg: &str,
-        backend: &mut Box<dyn PreferredSequencerDbBackend>,
-        shutdown_sender: &watch::Sender<()>,
-    ) {
-        if cfg!(debug_assertions) {
-            match backend.read_in_progress_batch().await {
-                Ok(None) => {}
-                other => {
-                    tracing::error!("{msg}: {other:?}");
-                    exit_rollup(shutdown_sender).await;
-                }
-            }
-        }
+/// Factory function to create the appropriate database implementation.
+pub(crate) async fn create_preferred_sequencer_db(
+    shutdown_sender: watch::Sender<()>,
+    is_replica: bool,
+    storage_path: &Path,
+    postgres_config: &Option<PostgresConfig>,
+) -> Result<Box<dyn Db>> {
+    if is_replica {
+        Ok(Box::new(ReplicaSequencerDb::new(shutdown_sender)))
```

Comment:
I get the idea, but this makes the pattern for handling the replica different from `BlobSender and CacheWarmUpExecutor`. We need a uniform approach, since all three components will go through the same state transition when a replica becomes primary. To reduce complexity, the code should handle this consistently across components. `Let’s bring back backend: Option<Box<dyn PreferredSequencerDbBackend>>.`